[FEATURE] Integrate media player from slub_web_sachsendigital

fschoelzel opened the pull request:
This PR integrates the video/audio player from slub_web_sachsendigital into a new MediaPlayer plugin. 
The PR is based on https://github.com/kitodo/kitodo-presentation/pull/796 and https://github.com/kitodo/kitodo-presentation/pull/818 from @dvoracek-slub 


> For testing this within the DFG Viewer, I have prepared a DDEV system at https://github.com/dvoracek-slub/ddev-dfgviewer-dist/tree/mediaplayer. It relies on https://github.com/dvoracek-slub/dfg-viewer/tree/dev-mediaplayer and includes a demo METS video document.
> 
> There also is a work-in-progress section in the documentation that describes some features and technical aspects of the player. `composer docs:serve` can be used for local preview.
> 
> Update (2022-07-06). Most notable changes:
> 
>     * Kitodo in general
>       
>       * Read production place and person names in `Mods` format class. I use this to test GND linking of some audio-related metadata.
>       * Add option to show full TOC.
> 
>     * Media player
>       
>       * It is now possible to set and export markers/segments in the media. The list is shown in audio mode (press F2).
>       * When the METS links preprocessed waveform data, this can be shown (and be used to edit markers). See the DDEV system for an example.
>       * First version of handling recordings that are split across multiple media files. Currently, this only works via page reloads.

Commit:
Add Webpack Dev Server

Commit:
(WIP) Add flexforms for player configuration

Commit:
(WIP) Start documenting media player

Commit:
Media Player: Implement equalizer

Commit:
Media Player: Integrate imported code

- Adopt build system from `slub_web_sachsendigital`
- Use player in plugin
- In Table of Contents, link to timecodes
- Make element ID configurable

Commit:
Media Player: Add plugin boilerplate

Commit:
Media Player: Add file group, read METS chapters

Commit:
Media Player: Chapter and page integration

Commit:
Start implementing waveform view

- Use Peaks.js to render waveform
- Add WAVEFORM fileGrp for preprocessed waveform data
- Connect to markers / marker table
- Integrate into audio and video mode
- First documentation

Commit:
Refactor: Extract share buttons from bookmark modal

Commit:
Unify handling of placeholders

Commit:
Media Player: Import code from `slub_web_sachsendigital`

Based upon commit slub/slub_web_sachsendigital#6e68a11

Commit:
Media Player: Build JS

Commit:
Configure jsconfig path

Commit:
Media Player: Fix and add poster

Commit:
Fix: Allow sub-second/float timecode

Commit:
Read MODS production places to allow linking valueURI

Commit:
Fix: Treat HTML as plain text in translation

Commit:
Fix: Stop Escape propagation only when actually handled

Commit:
Fix how sections are disabled in HelpModal

Commit:
Make player controls configurable in HTML

Commit:
Minor refactor file handling in MediaPlayerController

Commit:
Parse passed time range, not only start

Commit:
Query file FPS, use as fallback in player

Commit:
Handle fullscreen-element as player-view in DlfMediaPlayer

Commit:
Add container wrapper around player, add `media-panel`

Commit:
Allow to share segment, improve bookmark modal

- Share segment from marker table, or share active segment
- Let user select whether to share segment or current time (or none)
- Make "Copy Link" button an actual link, so that it is easier to
  actually bookmark

Commit:
Start implementing marker/segment edit and list

- Keybindings to add markers and segments
- `DlfMediaPlugin` for custom components that use a player
- `MarkerTable` for listing, editing, exporting markers
- Some documentation

Commit:
Extract actual keybindings for audio/video mode

Commit:
MarkerTable: Generate timerange URL

Commit:
Add some more documentation

Commit:
TOC: Add option to render full table of contents

Commit:
Use leaf nodes in TOC as chapters (instead of checking TYPE)

Commit:
Multi-File: Collect all files referenced from chapter marker

Commit:
Multi-File: Parse/use pageNo attribute for page dropdown

Commit:
Multi-File: Extend timecode links to chapter links

This will allow to give more information needed for cross-file links.

Commit:
Refactor: Extract UrlGenerator (instead of free functions)

Commit:
Multi-File: Allow multiple media files via page reloads

Commit:
Read MODS personal names to allow linking valueURI

Commit:
Show sub-second timecode when FPS is not available

Commit:
Media Player: Improve keyboard handling

- Better handling of Shift modifiers (don't list it in keybindings.json,
  and don't show it in help modal when obvious)
- On Mac, use Option key, and display variants appropriately
- Add more common "?" key for help modal

Commit:
Media Player: Add playrate slider for fine control

Commit:
(WIP) Clean up when removing media player

Commit:
Media Player: Add autoplay option to custom tag

Commit:
Media Player: Set seek bar box-sizing explicitly

Previously, positioning of the seek bar relied on box-sizing coming from
`slub_digitalcollections`.

Commit:
Media Player: Allow multiple player translation files

Commit:
Extend documentation

Commit:
fixup "Add file group" [fileGrpVideo]

Commit:
Media Player: Fix update of undefined player props

Commit:
Remove obsolete (?) `ts-expect-error` directives

Commit:
Fix translation labels (remove excess square bracket)

Commit:
Production build

Commit:
Fix timecode mode visibility/selection in bookmark modal

fschoelzel committed:
Code Maintenance - Rename getTitledata to getToplevelMetadata - Simplify check for doc existence - Fix returns in controllers' main functions - Rename Doc class to AbstractDocument

fschoelzel committed:
add custom functions and description from @dvoracek-slub

fschoelzel committed:
Fix directory naming for FlexForms and JavaScript in Media Player

fschoelzel committed:
Update Javascript Packages and fixed deprecations

update to packages recent major versions - node 18 - shaka-player v4 - peaks v3 and fixed deprecations - new build

fschoelzel committed:
Improvement of getTimecode in MetsDocument class

Move $extConf to getTimecode Method according to changes from this PR: https://github.com/kitodo/kitodo-presentation/pull/1118

fschoelzel committed:
update npm Packages and new Mediaplayer Build

fschoelzel committed:
fixed typecheck Errors in waveform.js

fschoelzel committed:
Media Player: Add Tooltip to Playrate Input Slider

to show actual number of the ratechange

fschoelzel committed:
[Bug Fix] Update segment properties in WaveForm.js to avoid  duplicate id issue with spread operator

fschoelzel committed:
Media Player: Add Mastodon Share Button with Input Form

Enter Mastodon Instance URL in the Input Field and Share the Media-Link with Timecode on your Profile

fschoelzel committed:
Media Player: Remove Twitter Share Button

update Kitodo Documentation

fschoelzel committed:
fixed BookmarkModal Test

fschoelzel committed:
Media Player: Fixed Backend Mastodon Share Button

fschoelzel committed:
Media Player: Removed BigPlayButton from UI

added singleClickForPlayAndPause

fschoelzel committed:
Ignored Typescript Errors

set ingore on errors for private functions. Tests are working

fschoelzel committed:
Media Player: Remove Playbutton from Desktop View - show only in Mobile View

When working with single Frames the Playbutton is always on top of the Video and disrupt the workflow - In Mobile View we assume there is no detailed work with the AV-Tools

fschoelzel committed:
Media Player: Use Constants from Typoscript to Seek and Rewind

add seekStep Value from Typoscript to navigate.rewind and navigate.seek

fschoelzel committed:
Media Player: Make Poster Configurable

Whether or not showing the Poster Image, if given, until playback is first started

fschoelzel committed:
Media Player: show BigPlayButton only once after playback has started - on mobile always

fschoelzel committed:
update npm packages and build

fschoelzel committed:
[MAINTENANCE] Use nested extension configuration

see: https://github.com/kitodo/kitodo-presentation/pull/1150

fschoelzel committed:
fix and refactor MetsDocument.php after rebase

fschoelzel committed:
specify getTimecode function to use fileGrpVideo instead of fileGrp for retrieving timecode and file IDs

fschoelzel committed:
Refactor Mods.php using php-mods-reader to improve code structure and readability

fschoelzel committed:
Fix: Refactor Mods.php using php-mods-reader to improve code structure

fschoelzel committed:
Refactor TableOfContentsController.php to include year information in entryArray as in Master Branch

fschoelzel committed:
Refactor MetsDocument to match with Musical Sources Feature

fschoelzel committed:
Update npm dependencies and Build

fschoelzel committed:
Media Player: [FEATURE] Select "Remove timestamp" removed in Bookmark function on MarkerTable

fschoelzel committed:
Refactor ScreenshotModal.js to fix image filename with extended Metadata to add timecode and fps

fschoelzel committed:
Rename class name to bookmark-markertable-hidden in the ScreenshotModal

fschoelzel committed:
Fix regular expression to correctly anchor .css and .less file extensions in webpack.config

Updated the regular expression in the webpack configuration to ensure that both .css and .less file extensions are correctly anchored at the end of the string. This prevents misleading operator precedence and ensures only the intended files are processed.

fschoelzel committed:
Media Player: Fix PHPStan Error - PHPDoc comments in MediaPlayerController

fschoelzel committed:
Media Player: Fix PHPStan errors by adding type annotations and checks in MediaPlayerController

fschoelzel committed:
add "video_frame_rate" property in AudioVideoMDTest

fschoelzel committed:
Media Player: Fixup - Update return type annotation for findFiles method

fschoelzel committed:
Refactor import paths in JavaScript files to path aliases

fschoelzel committed:
feat: Add alias imports support in Babel and Webpack configuration

- Updated .babelrc to include module-resolver plugin for alias imports.
- Modified webpack.config.js to resolve alias paths for cleaner imports.
- Update ESLint configuration for improved import resolution and compatibility

This change improves the readability and maintainability of import statements by allowing the use of path aliases.

fschoelzel committed:
[MAINTENANCE] Replace chr(31) with pack('C', 31) #1207

Codacy warning about usage of chr()

fschoelzel committed:
Refactor: Reduce complexity of getVideoInfo method

sebastian-meyer commented:
You placed a lot of JavaScript files in `Resources/Private/JavaScript/`. Are you sure those shouldn't be put in `Resources/Public/JavaScript/` instead, because they need to be accessible by the client?

Also, please put all unchanged third-party libraries in their own sub-directories separate from your own JS files. There are several reasons for that:
- It's easier to update dependencies if you know which files are part of a third-party library and don't contain any custom changes.
- We can exclude third-party libraries from all static analysis checks, especially regarding their code styling. (This will most likely greatly reduce the number of Codacy issues currently reported!)

fschoelzel commented:
> You placed a lot of JavaScript files in `Resources/Private/JavaScript/`. Are you sure those shouldn't be put in `Resources/Public/JavaScript/` instead, because they need to be accessible by the client?

This is correct and was done on purpose. All the files in the `Resources/Private/JavaScript/` folder are written to a compressed file called `DlfMediaPlayer.js` during the build process using webpack. This and the associated .map and vendor file are the only files that the client needs when loading the player.
Additional information to code organization: [MediaPlayer/Developers.rst](https://github.com/fschoelzel/kitodo-presentation/blob/dev-mediaplayer/Documentation/Features/MediaPlayer/Developers.rst#code-organization)

fschoelzel commented:
> Also, please put all unchanged third-party libraries in their own sub-directories separate from your own JS files. There are several reasons for that:
>- It's easier to update dependencies if you know which files are part of a third-party library and don't contain any custom changes.
>
>- We can exclude third-party libraries from all static analysis checks, especially regarding their code styling. (This will most likely greatly reduce the number of Codacy issues currently reported!)
>

As far as i know this two third-party files were copied from the original repository and rewritten by @dvoracek-slub to meet the ES6 typescript requirements:
[`Resources/Private/JavaScript/DlfMediaPlayer/3rd-party/EventTarget.js`](https://github.com/fschoelzel/kitodo-presentation/blob/dev-mediaplayer/Resources/Private/JavaScript/DlfMediaPlayer/3rd-party/EventTarget.js)
[`Resources/Private/JavaScript/DlfMediaPlayer/3rd-party/VideoFrame.js`](https://github.com/fschoelzel/kitodo-presentation/blob/dev-mediaplayer/Resources/Private/JavaScript/DlfMediaPlayer/3rd-party/VideoFrame.js)

The only other file I found is:
[`Resources/Private/JavaScript/SlubMediaPlayer/components/equalizer/filters.js`](https://github.com/fschoelzel/kitodo-presentation/blob/dev-mediaplayer/Resources/Private/JavaScript/SlubMediaPlayer/components/equalizer/filters.js)

Which in my opinion should not easily be updated, like the other two files, without deeply testing the whole function of the Player.

Apart from that, all other  third-party dependencies are located in the `Build/node_modules` folder and get compressed into the `DlfMediaVendor.js` file

fschoelzel commented:
I agree that the files created after the build process should be stored in the `Resources/Private/JavaScript/DlfMediaPlayer` folder instead of just the JavaScript folder.

fschoelzel committed:
[FIX] Add path for shaka-player/ui to lessOptions in webpack.config

fix warning in webpack build: Skipped data-uri embedding of images/xxx.svg because file not found

fschoelzel committed:
[FEAT] Migrate from PageRenderer to AssetCollector for asset management

Replaced PageRenderer with AssetCollector in MediaPlayerController.php for adding CSS and JavaScript assets -
Feature TYPO3 v10.3:
https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/10.3/Feature-90522-IntroduceAssetCollector.html#changelog-Feature-90522-IntroduceAssetCollector

fschoelzel committed:
Refactor: Update file paths for DlfMediaPlayer JavaScript files

fschoelzel committed:
[STYLE] Update Buttons in BookmarkModal and ScreenshotModal

sebastian-meyer commented on .gitignore:
```diff
@@ -11,10 +11,13 @@
 /Build/Documentation/dbdocs/var
+/Build/node_modules
 /Build/Test/.env
+composer.lock
```
Please don't ignore the `composer.lock`! We want to deliver a known working state of dependencies.
```suggestion
```

sebastian-meyer commented on Classes/Common/MetsDocument.php:
```diff
@@ -435,6 +435,7 @@ private function getFiles(array &$details, ?SimpleXMLElement $filePointers): voi
             // Check if file has valid @USE attribute.
             if (!empty($fileUse[$fileId])) {
                 $details['files'][$fileUse[$fileId]] = $fileId;
+                $details['all_files'][$fileUse[$fileId]][] = $fileId;
```
I think this is highly confusing. Also, `$details['files']` not being an array holding multiple file identifiers may be a bug.

I'd prefer adapting `$details['files']` to hold all file identifiers instead of introducing a confusing new variable with just a slightly different use case. What do you think?

fschoelzel replied:
The Problem is, there could be more than one `fptr->area` Element for each Physical Element. like here:
```xml
<mets:structMap TYPE="PHYSICAL">
    <mets:div ID="PHYS_0000" TYPE="physSequence">
        ...
        <mets:div ID="PHYS_0002" ORDER="2" TYPE="track">
            <mets:fptr FILEID="FILE_0002_THUMBS"/>
            <mets:fptr FILEID="FILE_0000_WAVEFORM"/>
            <mets:fptr>
                <mets:area FILEID="FILE_0000_DEFAULT_MPD" BEGIN="00:06:04" BETYPE="TIME" EXTENT="00:00:00" EXTTYPE="TIME"></mets:area>
            </mets:fptr>
            <mets:fptr>
                <mets:area FILEID="FILE_0000_DEFAULT_HLS" BEGIN="00:06:04" BETYPE="TIME" EXTENT="00:00:00" EXTTYPE="TIME"></mets:area>
            </mets:fptr>
            <mets:fptr>
                <mets:area FILEID="FILE_0000_DEFAULT_MP4" BEGIN="00:06:04" BETYPE="TIME" EXTENT="00:00:00" EXTTYPE="TIME"></mets:area>
            </mets:fptr>
        </mets:div>
       ...
    </mets:div>
</mets:structMap>
```

This is one reason why Kajetan introduced the `['all_files']` Array.
Otherwise, it just process the last file of the iteration.

If i decline this, i had rewrite everything which works with `['files']` now, and make it more dimensional.
I think this would be better to refactor, if it is merged first.

beatrycze-volk replied:
I would say that for now we should just take it in as it works. Improvements with naming can be done later.

sebastian-meyer commented on Classes/Common/MetsDocument.php:
```diff
@@ -1468,6 +1530,7 @@ private function getFileRepresentation(string $id, SimpleXMLElement $physicalNod
             // Check if file has valid @USE attribute.
             if (!empty($fileUse[$fileId])) {
                 $this->physicalStructureInfo[$id]['files'][$fileUse[$fileId]] = $fileId;
+                $this->physicalStructureInfo[$id]['all_files'][$fileUse[$fileId]][] = $fileId;
```
See above. I'd prefer adapting `files` instead of introducing `all_files`.

sebastian-meyer commented on Classes/ViewHelpers/MediaPlayerConfigViewHelper.php:
```diff
@@ -0,0 +1,131 @@
+
+        $localizationFactory = GeneralUtility::makeInstance(LocalizationFactory::class);
+
+        // TODO: Wouldn't there be a TypoScript utility?
```
Please have a look at https://docs.typo3.org/other/typo3/view-helper-reference/main/en-us/Global/Translate.html
It's not recommended to implement your own translation function, try to instead use one provided by the TYPO3 framework.

fschoelzel replied:
This ViewHelper uses the same TYPO3 Localization Core Functions, as the Fluid ViewHelper. But this ViewHelper exists to combine the Translations for the Templates and the Mediaplayer and pass them to the Shaka Player's UI. So the Mediaplayer gets translated with the standard TYPO3 Language files. Because it is used in the Javascript, the Fluid-Concept is not applicable here.

sebastian-meyer commented on Configuration/FlexForms/TableOfContents.xml:
```diff
@@ -80,6 +80,16 @@
+                            <label>LLL:EXT:dlf/Resources/Private/Language/locallang_be.xlf:plugins.tableofcontents.flexform.showFull</label>
+                            <config>
+                                <type>check</type>
+                                <default>0</default>
```
The default here should definitely be `1` in order to keep the current behavior! Please also make sure that this is the default in `TableOfContentsController` if the option was not set at all (i. e. after upgrading).

fschoelzel replied:
In this case, rendering the child entries, this is an added option to show all the child entries, which is not the current behavior. So set it to zero or not set it at all, keeps the default behavior.

see commit: 31955dd and added v12 compatibility: 1c278184579be7aa8f37226fd3155c3ce2e59028

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/Markers.test.js:
```diff
@@ -0,0 +1,70 @@
+          },
+        ],
+      },
+    })
```
This should fix CodeQL issue.
```suggestion
    });
```

sebastian-meyer commented on Build/.eslintrc.compat.js:
```diff
@@ -0,0 +1,20 @@
+  },
+  "ignorePatterns": [
+    // avoid Error: ELOOP: too many symbolic links encountered
+    "../Resources/Private/DevServer/Resources",
```
Please remove the DevServer from production code as it is not needed. Instead please add documentation on how to use DevServer for development or provide scripts for installing in a development environment.

fschoelzel replied:
I moved the Webpack DevServer to `./Build/Webpack/DevServer`, so it is not in the `./Resources/Private` Folder anymore and so more conform to TYPO3 Path Structure. Now it is more clear that it is for Testing next to the PHP Test Folders and the `webpack.config.js` fle.

fschoelzel replied:
a documentation how to use the webpack-dev-server is described in this Part:
https://github.com/fschoelzel/kitodo-presentation/blob/dev-mediaplayer/Documentation/Features/MediaPlayer/Developers.rst#webpack-dev-server

sebastian-meyer commented on Build/package.json:
```diff
@@ -0,0 +1,103 @@
+{
+  "name": "kitodo-presentation",
+  "version": "5.0.4",
```
This will be part of v6.0.0.
```suggestion
  "version": "6.0.0",
```

sebastian-meyer commented on Build/package.json:
```diff
@@ -0,0 +1,103 @@
+  },
+  "license": "GPL-3.0-or-later",
+  "homepage": "https://www.kitodo.org",
+  "devDependencies": {
```
Could the DevServer maybe added as a `devDependency` here?

fschoelzel replied:
"webpack-dev-server" was already in "devDependencies": {}

sebastian-meyer commented on Classes/Common/MetsDocument.php:
```diff
@@ -1537,6 +1592,63 @@ private function getPhysicalElements(array $elementNodes, array $fileUse): array
+        if (!empty($fileUse[$fileId])) {
+            $this->physicalStructureInfo[$elements[$order]]['files'][$fileUse[$fileId]] = $fileId;
+            // List all files of the fileGrp that are referenced on the page, not only the last one
+            $this->physicalStructureInfo[$elements[$order]]['all_files'][$fileUse[$fileId]][] = $fileId;
```
See above. I'd prefer adapting `files` instead of introducing `all_files`.

sebastian-meyer commented on Classes/Common/MetsDocument.php:
```diff
@@ -1537,6 +1592,63 @@ private function getPhysicalElements(array $elementNodes, array $fileUse): array
+
+        if (!empty($fileUse[$fileId])) {
+            $physInfo['files'][$fileUse[$fileId]] = $fileId;
+            $physInfo['all_files'][$fileUse[$fileId]][] = $fileId;
```
See above. I'd prefer adapting `files` instead of introducing `all_files`.

sebastian-meyer commented on Classes/Controller/TableOfContentsController.php:
```diff
@@ -123,6 +123,13 @@ private function getMenuEntry(array $entry, bool $recursive = false): array
         $entryArray['title'] = $this->setTitle($entry);
         $entryArray['volume'] = $entry['volume'];
+        if ($entry['videoChapter'] !== null) {
+            // Now consumers such as `slub_digitalcollections` may intercept clicks on these links
+            // and use the timecode to directly jump to that video position
```
Why isn't this implemented in Kitodo.Presentation directly? Do you really need additional code in another extension (like `slub_digitalcollections`) in order to jump to a video timecode position?

fschoelzel replied:
For choosing the Videochapter, the TOC is overwritten by javascript directly in Kitodo.Presentation. So it is implemented in Kitodo. The hint refers to the custom Kitodo TableOfContents Partials in `dfg-viewer` and `slub_digitalcollections`, which adds timecode information to TOC links at `section="{child.section}"`. Which is also present in Kitodo. So nothing to worry here.

sebastian-meyer commented on Classes/Format/Mods.php:
```diff
@@ -332,6 +334,76 @@ private function getPlaces(): void
+                            }
+                        }
+                }
+        }
```
Please fix indentation.

sebastian-meyer commented on Classes/Format/Mods.php:
In general, we shouldn't add more of these very specific metadata queries which concatenate multiple information about the same entry. Instead, the new feature for multi-level metadata should be used, which allows to configure those as grouped XPath queries.
Unfortunately, the feature is currently bugged, but hopefully this gets fixed soon.

fschoelzel replied:
This is something i would maintain and refactor in the Future. As it works for now.
For Example in a rewrite of the extractMetadata() Function.

issue: https://github.com/kitodo/kitodo-presentation/issues/1528

sebastian-meyer commented on Configuration/FlexForms/MediaPlayer.xml:
```diff
@@ -0,0 +1,163 @@
+                                <default>EXT:dlf/Resources/Private/Language/locallang_media.xlf</default>
+                            </config>
+                        </TCEforms>
+                    </settings.playerTranslations.baseFile>
```
Why make the translation file configurable? Using the TYPO3 localization functions you can already overwrite default translation in Typoscript.

fschoelzel replied:
The feature to config the translation file in the BE and in Typoscript has been removed: 0edde33a252ccbb01111ad7d8d444622e9eeb6e5
Now it just use the default translation file from /Resources

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/Markers.test.js:
```diff
@@ -0,0 +1,70 @@
+// @ts-check
+
+import { beforeEach, describe, expect, jest, test } from '@jest/globals';
```
```suggestion
import { describe, expect, jest, test } from '@jest/globals';
```
It seems like `beforeEach` is imported but never used.

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/components/waveform/WaveForm.js:
```diff
@@ -0,0 +1,392 @@
+
+import { clamp, e } from 'lib/util';
+import { action } from 'DlfMediaPlayer/lib/action';
+import DlfMediaPlayer from 'DlfMediaPlayer/DlfMediaPlayer';
```
`DlfMediaPlayer` seems to get imported but never used.

fschoelzel replied:
fixed with `@typedef import`

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/components/waveform/WaveForm.js:
```diff
@@ -0,0 +1,392 @@
+      return;
+    }
+
+    const { sample_rate, seconds_per_pixel, duration } = this.waveformdata;
+
+    this.minSamplesPerPixel = seconds_per_pixel * sample_rate;
+    this.maxSamplesPerPixel = clamp(
+      duration * sample_rate / this.$waveOverview.offsetWidth,
```
```suggestion
    const { sampleRate, secondsPerPixel, duration } = this.waveformdata;

    this.minSamplesPerPixel = secondsPerPixel * sampleRate;
    this.maxSamplesPerPixel = clamp(
      duration * sampleRate / this.$waveOverview.offsetWidth,
```
Please use `camelCase` for variables and properties.

fschoelzel replied:
f5243b944ef1f1b2b206e599b6e25050377f503a
revert variable renaming as it would break compatibility with waveform-data library

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/controls/FlatSeekBar.js:
```diff
@@ -0,0 +1,356 @@
+import shaka from 'shaka-player/dist/shaka-player.ui';
+
+import { e } from 'lib/util';
+import Chapters from 'DlfMediaPlayer/Chapters';
```
Again, this is imported but never used.

fschoelzel replied:
fixed with `@typedef import`

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/controls/FullScreenButton.js:
```diff
@@ -0,0 +1,68 @@
+// @ts-check
+
+import shaka from 'shaka-player/dist/shaka-player.ui';
```
Imported but never used.

fschoelzel replied:
fixed with `@typedef import`

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/controls/VideoTrackSelection.js:
```diff
@@ -0,0 +1,155 @@
+import shaka from 'shaka-player/dist/shaka-player.ui';
+
+import { e } from 'lib/util';
+import VariantGroups from 'DlfMediaPlayer/VariantGroups';
```
You know the drill... ;o)

fschoelzel replied:
fixed with `@typedef import`

sebastian-meyer commented on Resources/Private/JavaScript/DlfMediaPlayer/frontend/ShakaFrontend.js:
```diff
@@ -0,0 +1,551 @@
+import shaka from 'shaka-player/dist/shaka-player.ui';
+import 'shaka-player/ui/controls.less';
+
+import EventManager from 'lib/EventManager';
```
And again... I'll stop marking every single instance of imported but never used code. Please have a look at CodeQL which reports them as well.

fschoelzel replied:
fixed with `@typedef import`

sebastian-meyer commented on Resources/Private/JavaScript/SlubMediaPlayer/SlubMediaPlayer.js:
```diff
@@ -0,0 +1,572 @@
+      const value = el.getAttribute('value');
+
+      if (!key || !value) {
+        console.warn('Ignoring invalid <dlf-meta>');
```
Please remove debugging code.

sebastian-meyer commented on Resources/Private/JavaScript/SlubMediaPlayer/__snapshots__/Screenshot.test.js.snap:
What is this file? Is this part of a test or just remnants of development?

fschoelzel replied:
I had to find this out first. And copilot says:

Jest Snapshots sind ein mächtiges Werkzeug für das Testen von UI-Komponenten oder Datenstrukturen:
- Sie speichern eine serialisierte Version des erwarteten Ausgabezustands
- Bei jedem Testlauf wird die aktuelle Ausgabe mit dem gespeicherten Snapshot verglichen

### Praktische Bedeutung
Diese Datei wird:
- Automatisch von Jest generiert
- Sollte unter Versionskontrolle stehen
- Hilft dabei, unbeabsichtigte Änderungen am Verhalten der Anwendung zu erkennen
- Die Datei sollte nicht manuell editiert werden

fschoelzel replied:
As it is Part of the Jest Testing Framework, I would leave it there.

sebastian-meyer commented on Resources/Private/JavaScript/SlubMediaPlayer/components/equalizer/FrequencyResponse.js:
```diff
@@ -0,0 +1,134 @@
+        // Manual IIR filter
+        // https://www.earlevel.com/main/2016/12/01/evaluating-filter-frequency-response/
+        for (let i = 0; i < this.frequencies_.length; i++) {
+          const f = /** @type {number} */(this.frequencies_[i])
```
```suggestion
          const f = /** @type {number} */(this.frequencies_[i]);
```

sebastian-meyer commented on Resources/Private/JavaScript/SlubMediaPlayer/components/equalizer/filtersets/RiaaEq.js:
```diff
@@ -0,0 +1,355 @@
+   * @private
+   */
+  makeMidFilter() {
+    // midGain = 10 * Math.log10(1 + (baseBoost / freq) ** 2) - 10 * Math.log10(1 + (baseBoostRolloff / freq) ** 2)
+    // midGain / 10 = Math.log10(1 + (baseBoost / freq) ** 2) - Math.log10(1 + (baseBoostRolloff / freq) ** 2)
+    // midGain / 10 = Math.log10((1 + (baseBoost / freq) ** 2) / (1 + (baseBoostRolloff / freq) ** 2))
+    // 10 ^ (midGain / 10) = (1 + (baseBoost / freq) ** 2) / (1 + (baseBoostRolloff / freq) ** 2)
+
+    // 10 ^ (((10 * Math.log10(1 + baseBoost ** 2) - 10 * Math.log10(1 + baseBoostRolloff ** 2)) / 2) / 10) = (1 + (baseBoost / freq) ** 2) / (1 + (baseBoostRolloff / freq) ** 2)
+    // 10 ^ ((Math.log10((1 + baseBoost ** 2) / (1 + baseBoostRolloff ** 2))) / 2) = (1 + (baseBoost / freq) ** 2) / (1 + (baseBoostRolloff / freq) ** 2)
+    // sqrt((1 + baseBoost ** 2) / (1 + baseBoostRolloff ** 2)) = (1 + (baseBoost / freq) ** 2) / (1 + (baseBoostRolloff / freq) ** 2)
+
+    // For midFreq, ask Wolfram Alpha to solve
```
?

fschoelzel replied:
? me too.
What are we going to do with this?
Is it worthy to keep? In the end this are just comments from the developer

sebastian-meyer commented on Resources/Private/Less/DlfMediaPlayer.less:
```diff
@@ -0,0 +1,8 @@
+@import "../../../../slub_digitalcollections/Resources/Private/Less/Helper/Variables.less";
```
Please don't add dependencies to `slub_digitalcollections`!

fschoelzel replied:
removed by 15f670012800029b6b7fb866a174c6feb50085da

sebastian-meyer commented on Tests/Functional/FunctionalTestCase.php:
```diff
@@ -138,7 +138,10 @@ protected function getDlfConfiguration()
-                'fileGrpAudio' => 'AUDIO'
+                'fileGrpAudio' => 'AUDIO',
+                'fileGrpVideo' => 'VIDEO,DEFAULT',
+                'fileGrpVideo' => 'VIDEO,DEFAULT,VIDEO_ALT1',
```
Duplicate use of the array key `fileGrpVideo`.

fschoelzel replied:
deca0c6c34ae450b1de35b13910080316ebc880a

sebastian-meyer commented on import-mediaplayer.sh:
Please remove this file.

fschoelzel replied:
removed by e63d89c14dccb741857c88def72b0f7bf3f57ab9

sebastian-meyer reviewed (commented):
Some general remarks in addition to the detailed comments below (as discussed in our call):
* Please remove the DevServer or make it a dev-only dependency if that's possible.
* Please try to merge DlfMediaPlayer, SlubMediaPlayer and lib/ as much as possible, since those are dependencies of each other anyway which won't work as stand-alone implementations.
** At the moment it is really difficult to tell apart what those do, where they are used and how they depend on each other.
** The best way of handling this would be to have just two directories: One containing the third-party Shaka player (and possibly other unchanged third-party libraries) and one containing all the Kitodo-specific code.
* Please add license and author information to every non-third-party file.

fschoelzel committed:
fixup: add fileGrpVideo and fileGrpWaveform to new useGroups method

add mediaplayer useGroups to AbstractDocument after rewrite in #1404

fschoelzel committed:
add TYPO3 v12 compatibility to MediaPlayerController

fschoelzel committed:
fixup - [FEATURE] Implement `UseGroupsConfiguration` class (#1422)

fschoelzel committed:
fixup PhysicalStructure after rebase

fschoelzel committed:
[REFAC] Update MediaPlayerController to use video and thumbnail use groups

fschoelzel committed:
[FIX] Fixing key binding for Space Bar translation and localization

because of Issue in JS - Uncaught TypeError: A message must be provided as a String or AST.

fschoelzel committed:
update node packages

fschoelzel committed:
Move Webpack DevServer to Build/ Folder and update configuration and resource paths

fschoelzel committed:
Update .gitignore

remove composer.lock

Co-authored-by: Sebastian Meyer <sebastian.meyer@opencultureconsulting.com>

fschoelzel committed:
Update Resources/Private/JavaScript/DlfMediaPlayer/Markers.test.js

Co-authored-by: Sebastian Meyer <sebastian.meyer@opencultureconsulting.com>

fschoelzel committed:
Update Resources/Private/JavaScript/DlfMediaPlayer/components/waveform/WaveForm.js

use camelCase for variables and properties

Co-authored-by: Sebastian Meyer <sebastian.meyer@opencultureconsulting.com>

fschoelzel committed:
Update Resources/Private/JavaScript/SlubMediaPlayer/components/equalizer/FrequencyResponse.js

add missing semicolon

Co-authored-by: Sebastian Meyer <sebastian.meyer@opencultureconsulting.com>

fschoelzel committed:
remove import of slub_digitalcollections less variables and add them static

fschoelzel committed:
removed script to Import the Mediaplayer from slub_web_sachsendigital

fschoelzel committed:
moved bookmark mastodon share button to kitodo-presentation

fschoelzel committed:
Media Player: Remove Bookmark Twitter Share Function

fschoelzel committed:
Mediaplayer: Remove duplicate of the array key useGroupsVideo in Test File

fschoelzel committed:
Revert changes in WaveForm.js

Reverts variable renaming as it would break compatibility with waveform-data library:
- Library expects snake_case properties i.e. 'sample_rate'
- TypeScript definitions enforce 'sample_rate' naming

fschoelzel committed:
fix indentation

fschoelzel committed:
fix: add isset check for showFull setting in TableOfContentsController

Adds proper existence check for showFull configuration value to prevent PHP undefined array key warning.

fschoelzel committed:
[REFACTOR] Remove option to select further Mediaplayer Translation Files and simplified the ViewHelper

fschoelzel committed:
add Mediaplayer Plugin Icon

fschoelzel committed:
Media Player - Allow to set language translation file in Extension configuration

fschoelzel committed:
fix: improve type safety and clean up imports in media player components

fschoelzel committed:
fix: implement fullscreen utility functions

causing eslint errors: document.fullscreenElement() and fullscreenEnabled() is not supported in iOS Safari 15.6-15.8

fschoelzel committed:
fix: enhance null check for videoChapter in TableOfContentsController

fschoelzel committed:
[REFACTOR] fptr processing logic in MetsDocument to previous Version

fschoelzel committed:
fixing codacy issues and update node packages

fschoelzel committed:
[MAINTENANCE] update MediaPlayerConfigViewHelper to use SiteLanguage

Remove Superglobal and add dependency injection for SiteLanguage request and update deprecated getTwoLetterIsoCode

fschoelzel committed:
fixing codacy issues

fschoelzel committed:
[FIX] refactor to use RenderingContext for request variable

fschoelzel committed:
[FIX] Media Player: fixing seek-bar seek-marker

fschoelzel committed:
fixing codacy import issues

fschoelzel committed:
[FIX] move click-event listener for modals closeAll()

fschoelzel committed:
[FIX] fixing locale retrieval for phpstan

fschoelzel committed:
[MAINTENANCE] add License Information - build js files - update node packages

fschoelzel commented:
Regarding to this: https://github.com/kitodo/kitodo-presentation/pull/1320#pullrequestreview-2322668303
> Some general remarks in addition to the detailed comments below (as discussed in our call):
> 
>     * Please remove the DevServer or make it a dev-only dependency if that's possible.
> 
The few files to Test the Media Player have been moved to `Build/Webpack/DevServer/`

>     * Please try to merge DlfMediaPlayer, SlubMediaPlayer and lib/ as much as possible, since those are dependencies of each other anyway which won't work as stand-alone implementations.
>       ** At the moment it is really difficult to tell apart what those do, where they are used and how they depend on each other.
>       ** The best way of handling this would be to have just two directories: One containing the third-party Shaka player (and possibly other unchanged third-party libraries) and one containing all the Kitodo-specific code.
> 
Because this is already hard enough to handle, i would not vote to merge this all together. The reason, why it was arranged this way is clarified in the Kitodo/Mediaplayer Documentation. One Solution to seperate it for future JS libaries in the `Private/JavaScript/` folder, could be another parent folder for the Mediaplayer. 
For example: 

- `Resources/Private/JavaScript/DlfMediaPlayer/`
  - `DlfMediaPlayer` or `MediaPlayer`
  - `SlubMediaPlayer`
  - `lib`

What do you think?


>     * Please add license and author information to every non-third-party file.
I added the Kitodo License Information to the Javascript files in the `Resources/Public/JavaScript/DlfMediaPlayer/` folder. Does it need to be in every `Private/` JS file? Which the user never see?

beatrycze-volk reviewed (approved):
After Codacy errors related to PHP files are solved, the PR could be merged. The JS Codacy errors are mostly irrelevant or false positives like missing declarations which in fact are not missing.

fschoelzel committed:
minor fixes in php files for codacy

sebastian-meyer committed:
Merge branch 'main' into dev-mediaplayer

sebastian-meyer reviewed (approved):
Wow, outstanding work! What an achievement, especially considering how many different developers already worked on this code!

The javascript with all its cross-dependencies is still mostly a big mess, but that's not your fault and we can fix that later. I'll merge it for now in order to finally get to a stable 6.0 release.

Again: thank you very much for this big contribution! Great job!
